### ISSUE TITLE
Reassociation is needlessly suppressed when unrelated FastMathFlags are off

### ISSUE THREAD

**Issue by wjristow** [llvm/llvm-project#56483]
Reassociation enabled with `-ffast-math` only remains enabled when _all_ the FastMathFlags (FMF) remain on.  Here is a simple C++ program that demonstrates it:

```cpp
bool foo(float a, float b, float c) {
  float d = a + b + c;
  float e = c + a + b;
  return d == e;
}
```

Compiling the above with `-O2 -ffast-math` results in code that overtly returns `true`, as expected/desired:

```console
$ clang++ -S -o - -emit-llvm -O2 -ffast-math test.cpp
    ...
entry:
  ret i1 true
    ...
$
```

Compiling with simply `-O2` results in code containing 4 additions and a comparison, which will sometimes return `true` and sometimes `false` (again, as expected/desired).

The problem is that when an unrelated FMF is turned off, then the reassociation is suppressed, and the additions and comparison are again generated.  For example, disabling the reciprocal transformation suppresses the reassociation and resulting optimization:

```console
$ clang++ -S -o - -emit-llvm -O2 -ffast-math -fno-reciprocal-math test.cpp
    ...
entry:
  %add = fadd reassoc nnan ninf nsz contract afn float %a, %b
  %add1 = fadd reassoc nnan ninf nsz contract afn float %add, %c
  %add2 = fadd reassoc nnan ninf nsz contract afn float %a, %c
  %add3 = fadd reassoc nnan ninf nsz contract afn float %add2, %b
  %cmp = fcmp reassoc nnan ninf nsz contract afn oeq float %add1, %add3
  ret i1 %cmp
    ...
$
```

There aren't any divisions in the test-case, so suppressing the reciprocal transformation shouldn't impact the generated code.  (Reassociation requires the `reassoc` and `nsz` FastMathFlags to be on, so the reassociation/optimization should be done in the above example.)

To be explicit, it isn't a problem with `-fno-reciprocal-math` -- setting other FMF to values other than their "plain `fast` versions" also suppresses the optimization.  As an example, the 'fast' setting of FMA is `-ffp-contract=fast`, and changing it to `-ffp-contract=on`, has the equivalent problem of needlessly suppressing the optimization:

```console
$ clang++ -S -o - -emit-llvm -O2 -ffast-math -ffp-contract=on test.cpp
    ...
entry:
  %add = fadd reassoc nnan ninf nsz arcp afn float %a, %b
  %add1 = fadd reassoc nnan ninf nsz arcp afn float %add, %c
  %add2 = fadd reassoc nnan ninf nsz arcp afn float %a, %c
  %add3 = fadd reassoc nnan ninf nsz arcp afn float %add2, %b
  %cmp = fcmp reassoc nnan ninf nsz arcp afn oeq float %add1, %add3
  ret i1 %cmp
    ...
$
```

I have a patch that I'll post shortly, to fix this.

**Comment by wjristow** [comment]
> Proposed fix at [D129523](https://reviews.llvm.org/D129523).
> 
> As I note there, splitting that patch into two seems like a sensible idea.

I did split off a piece of that as a prep/cleanup step: [D129612](https://reviews.llvm.org/D129612).
That cleanup portion has been committed at 230c8c56f21cfe4e23a24793f3137add1af1d1f0.

**Comment by wjristow** [comment]
> Proposed fix at [D129523](https://reviews.llvm.org/D129523).

Fix committed at c6507930493bf57d88bcb1f7d83da1b3c08ab02a.
